src/phpDocumentor/FilesystemFactory.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+use \LogicException;
+use League\Flysystem\Adapter\Local;
+use League\Flysystem\Filesystem;
+use League\Flysystem\MountManager;
+use \DI\Container;
+
+/**
+ * Class FilesystemFactory
+ */
+final class FilesystemFactory
+{
+    /** @var MountManager */
+    private $mountManager;
+
+    /** @var Container */
+    private $container;
+
+    /** @var Filesystem */
+    private $filesystem;
+
+    /**
+     * @param MountManager $mountManager
+     * @param Container $container
+     */
+    public function __construct(MountManager $mountManager, Container $container)
+    {
+        $this->mountManager = $mountManager;
+        $this->container = $container;
+    }
+
+    /**
+     * Returns a Filesystem instance based on the scheme of the provided Dsn
+     *
+     * @param Dsn $dsn
+     * @return Filesystem
+     */
+    public function create(Dsn $dsn)
+    {
+        $dsnId = spl_object_hash($dsn);
```

Comment:
Interesting approach on creating a key! It does have one downside that is a bit technical and theoretical:

The DSN is a Value Object and the concept behind Value Objects is that they are immutable and can be recreated at any time. Since spl_object_hash returns a different id if a different object is used this will potentially result in multiple cache entries to the same location.

When we look at what is identifying for a location (because each Filesystem object is unique for each location) we can state that the DSN string in the DSN object is unique because that doesn't change, even when another DSN object is used.